src/zeit/content/cp/automatic.py

```diff
@@ -67,116 +54,170 @@ def values(self):
-        if self._v_retrieved_content > seen:
-            self._v_try_to_retrieve_content = True
-        else:
-            self._v_try_to_retrieve_content = False
-        return content
+    @cachedproperty
+    def _content_query(self):
+        return zope.component.getAdapter(
+            self, zeit.content.cp.interfaces.IContentQuery,
+            name=self.automatic_type)
+
+    def select_modules(self, *interfaces):
+        for module in self.values():
+            if any([x.providedBy(module) for x in interfaces]):
+                yield module
+
+
+def pop_filter(items, predicate=None):
+    """Remove the first object from the list for which predicate returns True;
+    no predicate means no filtering.
+    """
+    for i, item in enumerate(items):
+        if predicate is None or predicate(item):
+            items.pop(i)
+            return item
+
+
+def is_lead_candidate(content):
+    metadata = zeit.cms.content.interfaces.ICommonMetadata(content, None)
+    if metadata is None:
+        return False
+    return metadata.lead_candidate
+
+
+class ContentQuery(grok.Adapter):
+
+    grok.context(zeit.content.cp.interfaces.IRenderedArea)
+    grok.implements(zeit.content.cp.interfaces.IContentQuery)
+    grok.baseclass()
+
+    def __init__(self, context):
+        self.context = context
+
+    def __call__(self):
+        return []
+
+    @property
+    def start(self):
+        return self.context.start
+
+    @property
+    def rows(self):
+        return self.context.count
+
+    @cachedproperty
+    def existing_teasers(self):
+        cp = zeit.content.cp.interfaces.ICenterPage(self.context)
+        result = set()
+        result.update(cp.teasered_content_above(self.context))
+        result.update(cp.manual_content_below(self.context))
+        return result
+
+
+class SolrContentQuery(ContentQuery):
+
+    grok.name('query')
+
+    FIELDS = ' '.join(zeit.find.search.DEFAULT_RESULT_FIELDS)
+
+    def __init__(self, context):
+        super(SolrContentQuery, self).__init__(context)
+        self.query_string = self.context.raw_query
+        self.order = self.context.raw_order
+
+    def __call__(self):
+        return self._query_solr(self.query_string, self.order)
 
     def _query_solr(self, query, sort_order):
         result = []
         try:
-            for item in zeit.find.search.search(
-                    query, sort_order=sort_order,
-                    start=self._v_retrieved_content,
-                    rows=self.count_to_replace_duplicates):
-                content = zeit.cms.interfaces.ICMSContent(
-                    item['uniqueId'], None)
+            solr = zope.component.getUtility(zeit.solr.interfaces.ISolr)
+            response = solr.search(
+                query, sort=sort_order,
+                start=self.start,
+                rows=self.rows,
+                fl=self.FIELDS,
+                fq=self.filter_query)
+            self.total_hits = response.hits
```

Comment:
Schon klar dass es um die Paginierung geht. Daher ja die Idee, einfach `self.total_hits = len(result)` weiter unten zu speichern. Damit hat Friedbert für die Paginierung auch tatsächlich die Ergebnismenge in der Hand. (Wobei, kann Friedbert dann nicht selbst `len(result)` aufrufen? Oder was macht Friedbert genau?)